src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/RefDictionary.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Collections.Extensions
+{
+    public sealed class RefDictionary<K, V> : IEnumerable<(K Key, V Value)>
+    {
+        private int[] buckets;
+        private Entry[] entries;
+        private int count;
+        private int hashBits = 18;
+        private int hashMask;
+
+        private struct Entry
+        {
+            public K key;
+            public V value;
+            public int next;
+        }
+
+        public RefDictionary()
+        {
+            buckets = new int[1 << hashBits];
+            entries = new Entry[1 << hashBits];
+            hashMask = (1 << hashBits) - 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(K key)
+        {
+            return key.GetHashCode() & hashMask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(K key)
+        {
+            return buckets[GetBucketIndex(key)] - 1;
+        }
+
+        public bool TryGetValue(K key, out V value)
+        {
+            Entry[] entries = this.entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (true)
+            {
+                if ((uint)entryIndex >= (uint)entries.Length)
+                {
+                    break;
+                }
+
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    value = entries[entryIndex].value;
```

Comment:
should we return the ref as well?

Reply:
Just thinking Get could be new TValue() and TryGet could be default. I'm thinking of TValue being a List<T> etc. Thinking of this being very useful for aggregation, groupby etc. Will put in some tests to demo.

Reply:
Any early Nucleotide numbers with this?

Reply:
I'm seeing 40% perf on updates. I'll put in a more accurate Nucleotide perf tests.